core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -625,29 +625,33 @@ protected Object visitInPredicate(InPredicate node, Object context)
                 return set.contains(value);
             }
 
-            boolean hasUnresolvedValue = value instanceof Expression;
             boolean hasNullValue = false;
             boolean found = false;
-            List<Object> values = new ArrayList<>(valueList.getValues().size());
-            List<Type> types = new ArrayList<>(valueList.getValues().size());
+            ImmutableList.Builder<Object> values = ImmutableList.builderWithExpectedSize(valueList.getValues().size());
+            ImmutableList.Builder<Type> types = ImmutableList.builderWithExpectedSize(valueList.getValues().size());
 
             ResolvedFunction equalsOperator = metadata.resolveOperator(session, OperatorType.EQUAL, types(node.getValue(), valueList));
             for (Expression expression : valueList.getValues()) {
+                // Determine whether expression is null without processing it
+                if (expressionIsNull(expression)) {
```

Comment:
It describes what it does, but it doesn't describe why is it better. The commit message doesn't describe it either.
So -- why the change?